### ISSUE TITLE
Research Spike: ESLint 9 compatibility

### ISSUE THREAD

**Issue by ChristianMurphy** [mdx-js/eslint-mdx#477]
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

ESLint 9 is changing the plugin API: https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/
There may be changes needed in ESLint-MDX

### Solution

Check to see if any changed/removed APIs are used, migrate if needed.

### Alternatives

* No change may be needed
* We could remain on eslint 8 for a while if the migration is challenging

**Comment by JounQin** [comment]
<img width="518" alt="image" src="https://github.com/mdx-js/eslint-mdx/assets/8336744/47b22685-e979-420f-a453-2078860b4052">

I don't see how to test eslint at this point.

**Comment by ChristianMurphy** [comment]
https://github.com/eslint/eslint/releases/tag/v9.0.0-alpha.0

**Comment by remcohaszing** [comment]
Since the new flat config no longer resolves files, and supports `eslint.config.mjs`, I think ESLint MDX integrations can switch to ESM.

**Comment by JounQin** [comment]
> Since the new flat config no longer resolves files, and supports `eslint.config.mjs`, I think ESLint MDX integrations can switch to ESM.

I'd like to keep dual. Legacy `.eslintrc` configs are still supported.

**Comment by ChristianMurphy** [comment]
Legacy configuration are deprecated https://eslint.org/blog/2023/12/eslint-v9.0.0-alpha.0-released/

There are far less quirks and hacks packaging as pure ESM, rather than dealing with dual packaging limitations.

**Comment by JounQin** [comment]
https://github.com/mdx-js/eslint-mdx/issues/496

@ChristianMurphy 

`worker.mjs` will bypass all the limitations.

They're deprecated but still supported. I don't want to break anyone's current workflow before they're ready including myself.

**Comment by tianyingchun** [comment]
What's the plan?

**Comment by karlhorky** [comment]
If you don't use Markdown, then the current version `eslint-plugin-mdx@3.1.5` appears to work ESLint v9 by copying [the `flat` config](https://github.com/mdx-js/eslint-mdx/blob/d6fc093fb32ab58fb226e8cf42ac77399b8a4758/packages/eslint-plugin-mdx/src/configs/flat.ts#L8-L26) which is exported:

`eslint.config.js` or `eslint.config.mjs`

```js
import eslintMdx from 'eslint-mdx';
import mdx from 'eslint-plugin-mdx';

/** @type {import('eslint').Linter.FlatConfig} */
const config = [
  {
    files: ['**/*.mdx'],
    languageOptions: {
      sourceType: 'module',
      ecmaVersion: 'latest',
      parser: eslintMdx,
      globals: {
        React: false,
      },
    },
    plugins: {
      mdx,
    },
    rules: {
      'mdx/remark': 'warn',
      'no-unused-expressions': 'error',
      'react/react-in-jsx-scope': 0,
    },
  },
];

export default config;
```

**Comment by tianyingchun** [comment]
i  have integrating mdx into a plugin it can works fine for eslint V9 
https://github.com/hyperse-io/eslint-config-hyperse

**Comment by tianyingchun** [comment]
what's plan for ESLint V9, i used it in local, it seems that everything works fine except we need to bump  `eslint-plugin-markdown` to latest version (`5.1.0`) for now it used   "eslint-plugin-markdown": "^3.0.1"?

**Comment by reduckted** [comment]
ESLint v8 is EOL in one month (https://eslint.org/version-support/). Would you consider dropping v8 support to make it easier to migrate this plugin to v9? 

If anyone really needs v8 support, they can continue to use the current version of this plugin.

**Comment by remcohaszing** [comment]
[`@eslint/markdown`](https://github.com/eslint/markdown) (formerly `eslint-plugin-markdown` [6.0.0](https://github.com/eslint/markdown/releases/tag/v6.0.0) is ESM only. I think it’s fair to drop CJS support in other ESLint plugins as well.

**Comment by tianyingchun** [comment]
it's time to drop cjs +1 :)

**Comment by terrymun** [comment]
I'm wondering what's the state of adding eslint 9 (and `@eslint/markdown`) compatibility? AFAIK `@eslint/markdown` is already at version 6 and the one used by `eslint-mdx` is three major versions behind.

**Comment by JounQin** [comment]
I believe it's working fine with ESLint v9 currently.

PR welcome to bump deps without dropping ESLint v8 for now except there are great features we want but incompatible.

**Comment by reduckted** [comment]
@JounQin What is the reason for wanting to keep ESLint V8 support?

**Comment by JounQin** [comment]
<img width="790" alt="Image" src="https://github.com/user-attachments/assets/2986c39a-3170-4302-83b9-62aa7365760c" />

<img width="1204" alt="Image" src="https://github.com/user-attachments/assets/04197df6-b8b4-431c-8eaf-2b1c4ec1255c" />

Clearly there are still 50%+ eslint v8 users.

**Comment by reduckted** [comment]
Can't those users keep using v3.1.5 of this package? ESLint V8 has been EOL for many months now, and it's not like they'd be missing out on new features since v3.1.5 was released 12 months ago.

I'm just thinking of ways to reduce the complexity and maintenance burden of trying to support ESLint v8 and v9 at the same time. 😄

**Comment by bodograumann** [comment]
Exactly, @reduckted . This all-or-nothing approach was also taken by the vue libraries. They simply released a new major version compatible only with eslint-9 and for eslint-8-support  you had to use the old versions.
I think at this point anything fixing anything hindering upgrade should be the priority. We had to drop eslint-mdx to be able to do that.

**Comment by terrymun** [comment]
I agree with @reduckted and @bodograumann that a major version release of eslint-mdx that exclusively supports `eslint@9` + `@eslint/markdown@6` is probably the best way forward.

Right now we have to ignore error messages when consuming this eslint plugin because it consumes an outdated version of eslint-markdown that does not support react19.

**Comment by JounQin** [comment]
> Right now we have to ignore error messages when consuming this eslint plugin because it consumes an outdated version of eslint-markdown that does not support react19.

Maybe this is something as I mentioned:

> except there are great features we want but incompatible.

@terrymun Do you have any reproduction of such issues?

**Comment by terrymun** [comment]
@JounQin Ah my bad, I mixed up react19 and eslint9.

When installing `eslint-plugin-mdx`, you'd get an unmeet peer dependency warning if you're using it in a stack that uses eslint 9:

![Image](https://github.com/user-attachments/assets/502b6b0c-b886-411d-9593-4081920c8ace)

**Comment by JounQin** [comment]
PR welcome to

1. bump `eslint-plugin-markdown` to v5.1.0 which supports both eslint v8 and v9
2. Considering https://github.com/mdx-js/eslint-mdx/issues/534#issuecomment-2002882419, we can copy related processor codes from `eslint-plugin-markdown` and drop it as dependency

**Comment by JounQin** [comment]
[v3.2.0](https://github.com/mdx-js/eslint-mdx/releases/tag/eslint-plugin-mdx%403.2.0) has been released with `eslint-plugin-markdown` dropped.

**Comment by FunnyHector** [comment]
Hi, could someone please point out a migration guide from 3.1.5 to 3.2.0 please? We have

```javascript
    {
      files: ["*.mdx"],
      extends: "plugin:mdx/recommended",
      plugins: ["markdown"],
      processor: "markdown/markdown",
      rules: {
        "import/prefer-default-export": "off",
        "import/no-extraneous-dependencies": "off",
        "jsx-a11y/anchor-has-content": "off",
        "jsx-a11y/no-noninteractive-tabindex": "off",
        "react/prop-types": "off",
      },
      settings: {
        "mdx/code-blocks": true,
      },
    },
```
And with markdown dropped, this is breaking. Thanks!

**Comment by JounQin** [comment]
[`v3.3.0`](https://github.com/mdx-js/eslint-mdx/releases/tag/eslint-plugin-mdx%403.3.0) uses ESM as primary now, while commonjs is still supported.
